cpp/src/comms/mpi/comms_mpi.cpp

```diff
@@ -0,0 +1,291 @@
+      }
+      else if (sizeof(value_t) == 2) {
+        return MPI_UINT16_T;
+      }
+      else if (sizeof(value_t) == 4) {
+        return MPI_UINT32_T;
+      }
+      else if (sizeof(value_t) == 8) {
+        return MPI_UINT64_T;
+      }
+      else {
+        CUGRAPH_FAIL("unsupported type");
+      }
+    }
+  }
+  else if(std::is_same<value_t, float>::value) {
+    return MPI_FLOAT;
+  }
+  else if(std::is_same<value_t, double>::value) {
+    return MPI_DOUBLE;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+
+template <typename value_t>
+constexpr ncclDataType_t get_nccl_type() {
+  if (std::is_integral<value_t>::value) {
+    if (std::is_signed<value_t>::value) {
+      if (sizeof(value_t) == 1) {
+        return ncclInt8;
+      }
+      else if (sizeof(value_t) == 4) {
+        return ncclInt32;
+      }
+      else if (sizeof(value_t) == 8) {
+        return ncclInt64;
+      }
+      else {
+        CUGRAPH_FAIL("unsupported type");
+      }
+    }
+    else {
+      if (sizeof(value_t) == 1) {
+        return ncclUint8;
+      }
+      else if (sizeof(value_t) == 4) {
+        return ncclUint32;
+      }
+      else if (sizeof(value_t) == 8) {
+        return ncclUint64;
+      }
+      else {
+        CUGRAPH_FAIL("unsupported type");
+      }
+    }
+  }
+  else if(std::is_same<value_t, float>::value) {
+    return ncclFloat32;
+  }
+  else if(std::is_same<value_t, double>::value) {
+    return ncclFloat64;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+
+constexpr MPI_Op get_mpi_reduce_op(ReduceOp reduce_op) {
+  if (reduce_op == ReduceOp::SUM) {
+    return MPI_SUM;
+  }
+  else if (reduce_op == ReduceOp::MAX) {
+    return MPI_MAX;
+  }
+  else if (reduce_op == ReduceOp::MIN) {
+    return MPI_MIN;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+
+constexpr ncclRedOp_t get_nccl_reduce_op(ReduceOp reduce_op) {
+  if (reduce_op == ReduceOp::SUM) {
+    return ncclSum;
+  }
+  else if (reduce_op == ReduceOp::MAX) {
+    return ncclMax;
+  }
+  else if (reduce_op == ReduceOp::MIN) {
+    return ncclMin;
+  }
+  else {
+    CUGRAPH_FAIL("unsupported type");
+  }
+}
+#endif
+
+Comm::Comm(int p) : _p{p} {
+#if ENABLE_OPG
+  // MPI
+  int flag{}, mpi_world_size;
+
+  MPI_TRY(MPI_Initialized(&flag));
+
+  if (flag == false) {
+    int provided{};
+    MPI_TRY(MPI_Init_thread(nullptr, nullptr, MPI_THREAD_MULTIPLE, &provided));
+    if (provided != MPI_THREAD_MULTIPLE) {
+      MPI_TRY(MPI_ERR_OTHER);
+    }
+    _finalize_mpi = true;
+  }
+
+  MPI_TRY(MPI_Comm_rank(MPI_COMM_WORLD, &_rank));
+  MPI_TRY(MPI_Comm_size(MPI_COMM_WORLD, &mpi_world_size));
+  CUGRAPH_EXPECTS( (_p == mpi_world_size), 
+                   "Invalid input arguments: p should match the number of MPI processes.");
+
+  _mpi_comm = MPI_COMM_WORLD;
+
+  // CUDA
+
+  CUDA_TRY(cudaGetDeviceCount(&_device_count));
+  _device_id = _rank % _device_count; // FixMe : assumes each node has the same number of GPUs
```

Comment:
This seems like it could be an issues.  We should discuss configuration and how to handle
you also only set one device and not one per rank

Reply:
> This seems like it could be an issues. We should discuss configuration and how to handle

I think it is fine. We expect python to set the device and this code to be removed. This is just for C++ smoke tests for now so that I can validate the progress on the C++ part.

> you also only set one device and not one per rank

No, it does set one device per rank. In an OPG environment, this section of the code is traversed by all ranks, each one will retrieve and set its own GPU based on its rank.